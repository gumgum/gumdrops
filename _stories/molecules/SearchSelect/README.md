# SearchSelect

`<SearchSelect>` allows for external options to be provided and selected through a dropdown menu, such as a text input that calls an API endpoint for autocomplete options.

| Name         | Default           | Description                                                                                      | type     |
| ------------ | ----------------- | ------------------------------------------------------------------------------------------------ | -------- |
| debounceTime | 300               | Time to wait before the `onChange` handler is called                                             | Number   |
| isFocused    | false             | Sets the `autoFocus` input attribute to `true`, so the element will be focused on mount          | Bool     |
| isLoading    | false             | Indicate if the the loading component should render                                              | Bool     |
| onChange     |                   | Debounced change event handler that will be called with the current value in the input           | Function |
| onSelect     |                   | Select even handler which will be called with the selected option, e.g. `{ id: 1, title: 'foo'}` | Function |
| options      | []                | Array of options with the following shape `[{ id: 1, title: 'foo'}, { id: 2, title: 'bar'}]`     | Array    |
| placeholder  | 'Start Typing...' | Placeholder value to display on an empty input                                                   | String   |
| renderLoader |                   | Function to render a loading component such as `() => <LoadingDots size="sm" />`                 | Function |
| size         | 'sm'              | Size of the input, one of `xs`, `sm`, or `md`                                                    | String   |
