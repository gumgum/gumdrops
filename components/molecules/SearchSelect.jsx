import React, { Component } from 'react';
import cx from 'classnames';
import PropTypes from 'prop-types';
import SearchSelectOptions from './SearchSelectOptions';
import debounce from '../utils/debounce';

class SearchSelect extends Component {
    static propTypes = {
        debounceTime: PropTypes.number,
        isFocused: PropTypes.bool,
        isLoading: PropTypes.bool,
        onChange: PropTypes.func.isRequired,
        onSelect: PropTypes.func.isRequired,
        options: PropTypes.array,
        placeholder: PropTypes.string,
        renderLoader: PropTypes.func,
        size: PropTypes.oneOf(['xs', 'sm', 'md'])
    };

    static defaultProps = {
        size: 'sm',
        placeholder: 'Search',
        options: [],
        isFocused: false,
        isLoading: false,
        debounceTime: 300
    };

    state = {
        isOpen: false,
        searchTerm: ''
    };

    componentWillUnmount() {
        this._removeListeners();
    }

    _mapContainerRef = ref => (this._container = ref);

    _mapBtnRef = ref => (this._clearBtn = ref);

    _addListeners() {
        document.addEventListener('click', this._eventOutside);
        document.addEventListener('focus', this._eventOutside, true);
    }

    _removeListeners() {
        document.removeEventListener('click', this._eventOutside);
        document.removeEventListener('focus', this._eventOutside, true);
    }

    _eventOutside = ({ target }) => {
        if (!this._container.contains(target)) {
            this.setState({ isOpen: false });
            this._removeListeners();
        }
    };

    _showOnFocus = ({ target }) => {
        if (
            this.props.options.length &&
            this._container.contains(target) &&
            target !== this._clearBtn
        ) {
            this.setState({ isOpen: true });
            this._addListeners();
        }
    };

    handleOptionClick = item => {
        const { onSelect } = this.props;
        onSelect(item);
        this.setState({
            isOpen: false,
            searchTerm: item.title
        });
    };

    _clearSearch = event => {
        event.stopPropagation();
        this.setState({ searchTerm: '', isOpen: false });
        this._removeListeners();
    };

    _searchItem = ({ target: { value } }) => {
        const trimmedValue = value.trim();

        // Don't search if term is less than 2 characters
        if (trimmedValue !== '' && trimmedValue.length > 1) {
            this._callDebouncedSearch(trimmedValue);
        }

        this.setState({
            searchTerm: value,
            isOpen: true
        });
    };

    _callDebouncedSearch = debounce(value => {
        this.props.onChange(value);
    }, this.props.debounceTime);

    _getLabel() {
        const { placeholder } = this.props;
        const { searchTerm } = this.state;
        if (searchTerm) {
            return searchTerm;
        }
        return placeholder;
    }

    render() {
        const { placeholder, size, isLoading, renderLoader, options, isFocused } = this.props;
        const { isOpen, searchTerm } = this.state;

        return (
            <div
                aria-expanded={isOpen}
                aria-label={this._getLabel()}
                className={cx('gds-search-select', { 'gds-search-select--open': isOpen })}
                onFocus={this._showOnFocus}
                ref={this._mapContainerRef}
                role="listbox">
                <div className="gds-search-select__control">
                    <input
                        autoComplete="off"
                        autoFocus={isFocused}
                        className={cx('gds-form-group__text-input md', {
                            [`gds-form-group__text-input--${size}`]: size
                        })}
                        name="searchSelect"
                        onChange={this._searchItem}
                        placeholder={placeholder}
                        type="text"
                        value={searchTerm}
                    />
                    {searchTerm.length > 0 ? (
                        <button
                            aria-label="Clear search select"
                            name="clearSearchSelect"
                            onClick={this._clearSearch}
                            ref={this._mapBtnRef}
                            style={{
                                position: 'absolute',
                                padding: 0,
                                margin: 0,
                                top: '33%',
                                right: '0.6rem',
                                zIndex: 900,
                                display: 'inline-block',
                                border: 0,
                                borderRadius: 0
                            }}>
                            <i
                                className={cx('btl bt-times', {
                                    'bt-sm': size === 'xs',
                                    'bt-lg': size !== 'xs'
                                })}
                                style={{ display: 'inline-block' }}
                            />
                        </button>
                    ) : (
                        <i
                            className={cx('btl bt-search', {
                                'bt-lg': size === 'md',
                                'bt-sm': size === 'xs'
                            })}
                            style={{
                                position: 'absolute',
                                top: '31%',
                                right: '0.6rem',
                                zIndex: 900
                            }}
                        />
                    )}

                    {isLoading && renderLoader && renderLoader()}
                </div>
                <SearchSelectOptions options={options} handleOptionClick={this.handleOptionClick} />
            </div>
        );
    }
}

export default SearchSelect;
