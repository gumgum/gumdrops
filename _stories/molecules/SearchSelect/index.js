import React from 'react';
import { text, boolean, number } from '@storybook/addon-knobs';
import { optionalSelect } from '../../../components/utils/optionalSelect';
import { action } from '@storybook/addon-actions';
import LoadingDots from '../../../components/atoms/LoadingDots';

import readme from './README.md';
import SearchSelect from '../../../components/molecules/SearchSelect';

const options = [
    { id: 1, title: 'Lorem ipsum dolor sit amet' },
    { id: 2, title: 'Consectetur adipiscing elit ' },
    { id: 3, title: 'Sed do eiusmod tempor incididunt ut' },
    { id: 4, title: 'Labore et dolore magna aliqua' },
    { id: 5, title: 'Ut enim ad minim veniam' },
    { id: 6, title: 'Quis nostrud exercitation ullamco' },
    { id: 7, title: 'Laboris nisi ut aliquip ex ea commodo' },
    { id: 8, title: 'Consequat Duis' },
    { id: 9, title: 'Aute irure dolor in reprehenderit ' },
    { id: 10, title: 'In voluptate velit esse cillum dolore' },
    { id: 11, title: 'Eu fugiat nulla pariatur' },
    { id: 12, title: 'Deserunt mollit anim id est laborum' }
];

const sizeOptions = {
    xs: 'xs',
    sm: 'sm',
    md: 'md'
};

const component = () => (
    <SearchSelect
        placeholder={text('Placeholder', 'Start typing...')}
        options={options}
        onChange={action('ON_CHANGE')}
        onSelect={action('ON_SELECT')}
        size={optionalSelect('Size', sizeOptions, '')}
        isLoading={boolean('isLoading', false)}
        debounceTime={number('debounceTime', 300)}
        renderLoader={() => (
            <LoadingDots size="sm" style={{ left: 'auto', right: '3em', top: '45%' }} />
        )}
    />
);

export default [readme, component];
