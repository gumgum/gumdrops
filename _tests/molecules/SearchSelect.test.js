/* globals mount shallow */
import React from 'react';
import SearchSelect from '../../components/molecules/SearchSelect';

const options = [
    { id: 1, title: 'Lorem ipsum dolor sit amet' },
    { id: 2, title: 'Consectetur adipiscing elit ' },
    { id: 3, title: 'Sed do eiusmod tempor incididunt ut' },
    { id: 4, title: 'Labore et dolore magna aliqua' },
    { id: 5, title: 'Ut enim ad minim veniam' },
    { id: 6, title: 'Quis nostrud exercitation ullamco' },
    { id: 7, title: 'Laboris nisi ut aliquip ex ea commodo' },
    { id: 8, title: 'Consequat Duis' },
    { id: 9, title: 'Aute irure dolor in reprehenderit ' },
    { id: 10, title: 'In voluptate velit esse cillum dolore' },
    { id: 11, title: 'Eu fugiat nulla pariatur' },
    { id: 12, title: 'Deserunt mollit anim id est laborum' }
];

const defaultProps = {
    onSelect: () => {},
    onChange: () => {}
};

const addEventListener = global.document.addEventListener;
const removeEventListener = global.document.removeEventListener;

describe('Expect <SearchSelect>', () => {
    it('to render', () => {
        const wrapper = shallow(<SearchSelect options={options} {...defaultProps} />);
        expect(wrapper).toMatchSnapshot();
    });

    it('to render a loading component', () => {
        const Loader = jest.fn();
        mount(<SearchSelect options={options} {...defaultProps} isLoading renderLoader={Loader} />);
        expect(Loader).toBeCalled();
    });

    it('to add event listeners to the document and open when focused', () => {
        global.document.addEventListener = jest.fn();
        const wrapper = mount(<SearchSelect options={options} {...defaultProps} />);
        const input = wrapper.find('input[name="searchSelect"]').first();
        input.simulate('focus');
        expect(global.document.addEventListener).toBeCalledTimes(2);
        expect(global.document.addEventListener.mock.calls[0][0]).toBe('click');
        expect(global.document.addEventListener.mock.calls[1][0]).toBe('focus');
        expect(wrapper.state().isOpen).toBe(true);
        global.document.addEventListener = addEventListener;
    });

    it('to remove event listeners and close when capturing outside events', () => {
        global.document.removeEventListener = jest.fn();
        const wrapper = mount(<SearchSelect options={options} {...defaultProps} />);
        const instance = wrapper.instance();
        instance._eventOutside({ target: document.createElement('div') });
        expect(global.document.removeEventListener).toBeCalledTimes(2);
        expect(global.document.removeEventListener.mock.calls[0][0]).toBe('click');
        expect(global.document.removeEventListener.mock.calls[1][0]).toBe('focus');
        expect(wrapper.state().isOpen).toBe(false);
        global.document.removeEventListener = removeEventListener;
    });

    it('to remove event listeners and close when clearing selected options', () => {
        global.document.removeEventListener = jest.fn();
        const wrapper = mount(<SearchSelect options={options} {...defaultProps} />);

        const input = wrapper.find('input[name="searchSelect"]').first();
        input.simulate('focus');
        wrapper.setState({ searchTerm: 'search term' });
        expect(wrapper.state().isOpen).toBe(true);

        const clear = wrapper.find('button[name="clearSearchSelect"]').first();
        clear.simulate('click');

        expect(global.document.removeEventListener).toBeCalledTimes(2);
        expect(global.document.removeEventListener.mock.calls[0][0]).toBe('click');
        expect(global.document.removeEventListener.mock.calls[1][0]).toBe('focus');
        expect(wrapper.state().isOpen).toBe(false);
        global.document.removeEventListener = removeEventListener;
    });

    it('to remove event listeners and close when clearing selected options', () => {
        global.document.removeEventListener = jest.fn();
        const wrapper = mount(<SearchSelect options={options} {...defaultProps} />);

        const input = wrapper.find('input[name="searchSelect"]').first();
        input.simulate('focus');
        wrapper.setState({ searchTerm: 'search term' });
        expect(wrapper.state().isOpen).toBe(true);

        const clear = wrapper.find('button[name="clearSearchSelect"]').first();
        clear.simulate('click');

        expect(global.document.removeEventListener).toBeCalledTimes(2);
        expect(global.document.removeEventListener.mock.calls[0][0]).toBe('click');
        expect(global.document.removeEventListener.mock.calls[1][0]).toBe('focus');
        expect(wrapper.state().isOpen).toBe(false);
        global.document.removeEventListener = removeEventListener;
    });

    it('to call onSelect when handling a clicked option and update state', () => {
        const onSelect = jest.fn();
        const wrapper = mount(
            <SearchSelect options={options} {...defaultProps} onSelect={onSelect} />
        );
        wrapper.setState({ isOpen: false });
        const inst = wrapper.instance();
        inst.handleOptionClick(options[0]);
        expect(onSelect).toBeCalledWith(options[0]);
        expect(wrapper.state()).toEqual({ isOpen: false, searchTerm: options[0].title });
    });

    it('to call onChange after debounce', () => {
        jest.useFakeTimers();

        const onChange = jest.fn();
        const wrapper = mount(
            <SearchSelect options={options} {...defaultProps} onChange={onChange} />
        );
        const input = wrapper.find('input[name="searchSelect"]').first();
        input.simulate('change', { target: { value: options[0].title } });
        expect(onChange).not.toBeCalled();
        jest.runAllTimers();
        expect(onChange).toBeCalledWith(options[0].title);
    });

    it('to add classes based on size prop', () => {
        const wrapper1 = shallow(<SearchSelect options={options} {...defaultProps} size="xs" />);
        expect(wrapper1).toMatchSnapshot();

        const wrapper2 = shallow(<SearchSelect options={options} {...defaultProps} size="sm" />);
        expect(wrapper2).toMatchSnapshot();

        const wrapper3 = shallow(<SearchSelect options={options} {...defaultProps} size="md" />);
        expect(wrapper3).toMatchSnapshot();
    });
});
