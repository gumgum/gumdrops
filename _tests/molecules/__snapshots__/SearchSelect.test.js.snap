// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Expect <SearchSelect> to add classes based on size prop 1`] = `
<div
  aria-expanded={false}
  aria-label="Search"
  className="gds-search-select"
  onFocus={[Function]}
  role="listbox"
>
  <div
    className="gds-search-select__control"
  >
    <input
      autoComplete="off"
      autoFocus={false}
      className="gds-form-group__text-input md gds-form-group__text-input--xs"
      name="searchSelect"
      onChange={[Function]}
      placeholder="Search"
      type="text"
      value=""
    />
    <i
      className="btl bt-search bt-sm"
      style={
        Object {
          "position": "absolute",
          "right": "0.6rem",
          "top": "31%",
          "zIndex": 900,
        }
      }
    />
  </div>
  <SearchSelectOptions
    handleOptionClick={[Function]}
    noResultsMessage="No results found"
    options={
      Array [
        Object {
          "id": 1,
          "title": "Lorem ipsum dolor sit amet",
        },
        Object {
          "id": 2,
          "title": "Consectetur adipiscing elit ",
        },
        Object {
          "id": 3,
          "title": "Sed do eiusmod tempor incididunt ut",
        },
        Object {
          "id": 4,
          "title": "Labore et dolore magna aliqua",
        },
        Object {
          "id": 5,
          "title": "Ut enim ad minim veniam",
        },
        Object {
          "id": 6,
          "title": "Quis nostrud exercitation ullamco",
        },
        Object {
          "id": 7,
          "title": "Laboris nisi ut aliquip ex ea commodo",
        },
        Object {
          "id": 8,
          "title": "Consequat Duis",
        },
        Object {
          "id": 9,
          "title": "Aute irure dolor in reprehenderit ",
        },
        Object {
          "id": 10,
          "title": "In voluptate velit esse cillum dolore",
        },
        Object {
          "id": 11,
          "title": "Eu fugiat nulla pariatur",
        },
        Object {
          "id": 12,
          "title": "Deserunt mollit anim id est laborum",
        },
      ]
    }
  />
</div>
`;

exports[`Expect <SearchSelect> to add classes based on size prop 2`] = `
<div
  aria-expanded={false}
  aria-label="Search"
  className="gds-search-select"
  onFocus={[Function]}
  role="listbox"
>
  <div
    className="gds-search-select__control"
  >
    <input
      autoComplete="off"
      autoFocus={false}
      className="gds-form-group__text-input md gds-form-group__text-input--sm"
      name="searchSelect"
      onChange={[Function]}
      placeholder="Search"
      type="text"
      value=""
    />
    <i
      className="btl bt-search"
      style={
        Object {
          "position": "absolute",
          "right": "0.6rem",
          "top": "31%",
          "zIndex": 900,
        }
      }
    />
  </div>
  <SearchSelectOptions
    handleOptionClick={[Function]}
    noResultsMessage="No results found"
    options={
      Array [
        Object {
          "id": 1,
          "title": "Lorem ipsum dolor sit amet",
        },
        Object {
          "id": 2,
          "title": "Consectetur adipiscing elit ",
        },
        Object {
          "id": 3,
          "title": "Sed do eiusmod tempor incididunt ut",
        },
        Object {
          "id": 4,
          "title": "Labore et dolore magna aliqua",
        },
        Object {
          "id": 5,
          "title": "Ut enim ad minim veniam",
        },
        Object {
          "id": 6,
          "title": "Quis nostrud exercitation ullamco",
        },
        Object {
          "id": 7,
          "title": "Laboris nisi ut aliquip ex ea commodo",
        },
        Object {
          "id": 8,
          "title": "Consequat Duis",
        },
        Object {
          "id": 9,
          "title": "Aute irure dolor in reprehenderit ",
        },
        Object {
          "id": 10,
          "title": "In voluptate velit esse cillum dolore",
        },
        Object {
          "id": 11,
          "title": "Eu fugiat nulla pariatur",
        },
        Object {
          "id": 12,
          "title": "Deserunt mollit anim id est laborum",
        },
      ]
    }
  />
</div>
`;

exports[`Expect <SearchSelect> to add classes based on size prop 3`] = `
<div
  aria-expanded={false}
  aria-label="Search"
  className="gds-search-select"
  onFocus={[Function]}
  role="listbox"
>
  <div
    className="gds-search-select__control"
  >
    <input
      autoComplete="off"
      autoFocus={false}
      className="gds-form-group__text-input md gds-form-group__text-input--md"
      name="searchSelect"
      onChange={[Function]}
      placeholder="Search"
      type="text"
      value=""
    />
    <i
      className="btl bt-search bt-lg"
      style={
        Object {
          "position": "absolute",
          "right": "0.6rem",
          "top": "31%",
          "zIndex": 900,
        }
      }
    />
  </div>
  <SearchSelectOptions
    handleOptionClick={[Function]}
    noResultsMessage="No results found"
    options={
      Array [
        Object {
          "id": 1,
          "title": "Lorem ipsum dolor sit amet",
        },
        Object {
          "id": 2,
          "title": "Consectetur adipiscing elit ",
        },
        Object {
          "id": 3,
          "title": "Sed do eiusmod tempor incididunt ut",
        },
        Object {
          "id": 4,
          "title": "Labore et dolore magna aliqua",
        },
        Object {
          "id": 5,
          "title": "Ut enim ad minim veniam",
        },
        Object {
          "id": 6,
          "title": "Quis nostrud exercitation ullamco",
        },
        Object {
          "id": 7,
          "title": "Laboris nisi ut aliquip ex ea commodo",
        },
        Object {
          "id": 8,
          "title": "Consequat Duis",
        },
        Object {
          "id": 9,
          "title": "Aute irure dolor in reprehenderit ",
        },
        Object {
          "id": 10,
          "title": "In voluptate velit esse cillum dolore",
        },
        Object {
          "id": 11,
          "title": "Eu fugiat nulla pariatur",
        },
        Object {
          "id": 12,
          "title": "Deserunt mollit anim id est laborum",
        },
      ]
    }
  />
</div>
`;

exports[`Expect <SearchSelect> to render 1`] = `
<div
  aria-expanded={false}
  aria-label="Search"
  className="gds-search-select"
  onFocus={[Function]}
  role="listbox"
>
  <div
    className="gds-search-select__control"
  >
    <input
      autoComplete="off"
      autoFocus={false}
      className="gds-form-group__text-input md gds-form-group__text-input--sm"
      name="searchSelect"
      onChange={[Function]}
      placeholder="Search"
      type="text"
      value=""
    />
    <i
      className="btl bt-search"
      style={
        Object {
          "position": "absolute",
          "right": "0.6rem",
          "top": "31%",
          "zIndex": 900,
        }
      }
    />
  </div>
  <SearchSelectOptions
    handleOptionClick={[Function]}
    noResultsMessage="No results found"
    options={
      Array [
        Object {
          "id": 1,
          "title": "Lorem ipsum dolor sit amet",
        },
        Object {
          "id": 2,
          "title": "Consectetur adipiscing elit ",
        },
        Object {
          "id": 3,
          "title": "Sed do eiusmod tempor incididunt ut",
        },
        Object {
          "id": 4,
          "title": "Labore et dolore magna aliqua",
        },
        Object {
          "id": 5,
          "title": "Ut enim ad minim veniam",
        },
        Object {
          "id": 6,
          "title": "Quis nostrud exercitation ullamco",
        },
        Object {
          "id": 7,
          "title": "Laboris nisi ut aliquip ex ea commodo",
        },
        Object {
          "id": 8,
          "title": "Consequat Duis",
        },
        Object {
          "id": 9,
          "title": "Aute irure dolor in reprehenderit ",
        },
        Object {
          "id": 10,
          "title": "In voluptate velit esse cillum dolore",
        },
        Object {
          "id": 11,
          "title": "Eu fugiat nulla pariatur",
        },
        Object {
          "id": 12,
          "title": "Deserunt mollit anim id est laborum",
        },
      ]
    }
  />
</div>
`;
